### PR TITLE
backend/feature : Special Zone Fare Policy & Tag

### PR DESCRIPTION
**Pull request by khuzema786**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description

**Area Field in Fare Product**
1. **Default** : It is for all the areas including special locations unless they have their own Area fare product in DB. (Mandatory for all set of vehicle variants and merchants)
2. **Pickup_SpecialLocationId** : It is for a pickup special location. (Optional)
3. **Drop_SpecialLocationId** : It is for a drop special location. (Optional)

**Special Location Tag Flow**
1. **BAP RideSearch to BPP Search CallAPI** -> Calculate the SpecialLocationTag and store it in SearchRequest, SpecialZoneQuotes & Estimates BPP tables.
2. **BPP OnSearchCallback to BAP OnSearch CallAPI** -> Store the Quotes & Estimates in BAP Table.
3. **BAP RideSearch Results** -> Send the specialLocationTag in the response from the Quotes or Estimates BAP tables.
4. **BAP Select to BPP Select** -> Send SpecialLocationTag from SearchRequest BPP table in nearbyRideRequest API and FCM Popup.
5. **BPP RespondQuote to BAP OnSelect** -> Store SpecialLocationTag in DriverQuote BPP table and pass it onto BAP so that it can also store in Quotes BAP tables.
6. **BAP OnSelect** - Auto Assigned Flow -> Creates Booking BAP table entry using the lowest fare Quotes BAP table.
7. **BAP RideList** -> Send SpecialLocationTag from Booking BAP table.

**Rider App UI API Changes**
```
GET : {{baseUrl_app}}/rideSearch/:searchId/results
{ quotes : [{ .., specialLocationTag : Maybe Text }], (for OnDemandCab)
  estimates : [{ .., specialLocationTag : Maybe Text }]
}
```
```
GET : {{baseUrl_app}}/rideBooking/list
{ list : [{ .., specialLocationTag : Maybe Text }]
}
```
```
POST : {{baseUrl_app}}/rideBooking/:bookingId
{ .., specialLocationTag : Maybe Text }
```
**Driver App UI API Changes**
```
Driver Search Request PopUp (FCMData)
{ .., specialLocationTag : Maybe Text }
```
```
GET : {{baseURL_namma_P}}/driver/nearbyRideRequest
{ searchRequestsForDriver : [{ .., specialLocationTag : Maybe Text }]
}
```
```
GET : {{baseURL_namma_P}}/driver/ride/list
{ list : [{ .., specialLocationTag : Maybe Text }]
}
```

**Special Location Tags**
1. "SureMetro - PICKUP" | "SureMetro - DROP"
2. "UnSureMetro - PICKUP" | "UnSureMetro - DROP"
3. "SureAirport - PICKUP" | "SureAirport - DROP"
4. "UnSureAirport - PICKUP" | "UnSureAirport - DROP"
5. "SureSchool - PICKUP" | "SureSchool - DROP"
6. "UnSureSchool - PICKUP" | "UnSureSchool - DROP"
7. "SureHospital - PICKUP" | "SureHospital - DROP"
8. "UnSureHospital - PICKUP" | "UnSureHospital - DROP"
9. "SureStation - PICKUP" | "SureStation - DROP"
10. "UnSureStation - PICKUP" | "UnSureStation - DROP"

**Priority Logic**
Screenshot 2023-05-25 at 7 25 06 PM

1) If Pickup = Metro Station(pickupPriority= 2) & Drop = Hospital(dropPriority= 1), then, the mark the ride request as ‘Hospital Drop’ ride
2) If Pickup = Metro Station(pickupPriority= 2) & Drop = Airport(dropPriority= 4), then, the mark the ride request as ‘Metro Station Pickup’ ride
3) If Pickup = Metro Station(pickupPriority= 2) & Drop = Metro Station(dropPriority= 2), then, the mark the ride as ‘Metro Station Pickup’ ride
When pickup and drop zones have same priority, mark the ride with pickup location’s special zone
4) If the pickup location is a special zone(say Airport) but drop location is not a special zone, mark the ride as ‘Airport Pickup’ ride.

### Additional Changes

- [x] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [x] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FarePolicy.hs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    Just specialLocationLinks -> do
+      let mbFlowType = listToMaybe $ nub $ (.flow) <$> specialLocationLinks
+      return $ fromMaybe SpecialLocationLink.NORMAL mbFlowType
+    Nothing -> return SpecialLocationLink.NORMAL
+
+getFarePolicyForVariant :: (MonadFlow m, MonadReader r m, HasEsqEnv m r, CacheFlow m r) => Id Merchant -> LatLong -> LatLong -> Variant -> m (Maybe FarePolicy)
+getFarePolicyForVariant merchantId fromlocaton toLocation vehVariant = do
+  mbSpecialLocationLink <- SpecialLocation.getSpecialLocationLinkForVariant merchantId.getId fromlocaton toLocation vehVariant
+  case mbSpecialLocationLink of
+    Just specialLocationLink -> QFP.findById (Id specialLocationLink.farePolicyId)
```
Comment (DonDokidon) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FarePolicy.hs:30]:
```
1. rename FarePolicyDetails to FarePolicyTypeDetails
2. create new 
data FarePolicyTrait = NormalTrait FarePolicyTypeDetails | SpecialZonesTrait FarePolicySpecialZoneTrait
data FarePolicySpecialZoneTrait = FarePolicySpecialZoneTrait {
  -- there you implement all your task fields and necessaries like pickupPriority and other stuff, for each pickupPriority you'll have its own FarePolicyTypeDetails
}
data FarePolicyTraitsListItem = FarePolicyTraitListItem {
   farePolicyTrait :: FarePolicyTrait,
   traitPriority :: Int
}
3. replace FarePolicy.farePolicyDetails with FarePolicy.farePolicyTraitsList (NonEmpty, always have item with priority 0, which is most likely just NormalTrait)

Now when you passing ride info to count fareParams, you have to process this list from highest priority number to 0 (which is default case, which can handle just everything) and choose Trait with highest priority (currently we will have only 0 and 1 priorities, where 1 for FarePolicySpecialZoneTrait, but potentially we can create more easy enough) and count FareParams based on choosen trait. 
```
That's how i proposed to implement this and i as far as i understand your way is far from proposed approach. Did you find some issues with this? Or it was just too hard to implement?
@hkmangla I'd like to implement it that way, do we have some time before merging this, so i could refactor this PR?

Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FarePolicy.hs:30]: @DonDokidon Yes, but with this approach for 72 Special Locations will have 72 different fare policies even though if they are same, it would get hard to maintain right ?

Comment (DonDokidon) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FarePolicy.hs:30]: No? You will have only one FarePolicy, which will contain only one `FPSpecZoneTrait`, which will have 72 different `FPSpecZoneTraitZoneCosts` (idk why it's 72 though, you do care only on highest priority, so you'll need only 10 of ZoneCosts entities, 5 for highest pickup. 5 for highest drops). And even if you wanna have one single `FPSpecZoneTraitZoneCosts` for all 10 types of pickup/drop priority, you can make that with this approach, but IMO THAT will be harder to maintain. It will be easy to add directly from DB, by creating one `FPSpecZoneTraitZoneCosts` and assigning that to all 10 pickup/drops, using id ref. But how will you do it, when you will set it via dashboard? Will you check each pickup/drop if their settings are exactly the same and will create one single `FPSpecZoneTraitZoneCosts` entity for them in that case?

File: Backend/dev/migrations/dynamic-offer-driver-app/0164-special-location-tag.sql [hunk 2]
```diff
@@ -0,0 +1,140 @@
+ALTER TABLE atlas_driver_offer_bpp.quote_special_zone ADD column special_location_tag text;
+ALTER TABLE atlas_driver_offer_bpp.estimate ADD column special_location_tag text;
+ALTER TABLE atlas_driver_offer_bpp.driver_quote ADD column special_location_tag text;
+ALTER TABLE atlas_driver_offer_bpp.search_request ADD column special_location_tag text;
+ALTER TABLE atlas_driver_offer_bpp.fare_policy ADD column "description" text;
+CREATE TABLE atlas_driver_offer_bpp.special_location_priority (
```
Comment (hkmangla) [inline Backend/dev/migrations/dynamic-offer-driver-app/0164-special-location-tag.sql:6]: Add proper INDEXES in DB as per the queries being used.

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareProduct.hs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import Tools.Error
+
+getPickupSpecialLocation ::
+  (MonadFlow m, MonadReader r m, HasEsqEnv m r) =>
+  Id Merchant ->
+  DSpecialLocation.SpecialLocation ->
+  m (DSpecialLocation.SpecialLocation, Int)
+getPickupSpecialLocation merchantId pickupSpecialLocation = do
+  pickupSpecialLocationPriority <- QSpecialLocationPriority.findByMerchantIdAndCategory merchantId.getId pickupSpecialLocation.category
+  return (pickupSpecialLocation, maybe 0 (.pickupPriority) pickupSpecialLocationPriority)
```
Comment (0utkarsh) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareProduct.hs:33]: Why is the priority highest by default? Will this case never arise?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/FarePolicy.hs [hunk 4]
```diff
@@ -74,19 +72,19 @@ validateUpdateFarePolicyRequest UpdateFarePolicyReq {..} =
...
+--     ListFarePolicyRes
+--       { oneWayFarePolicies = map makeFarePolicyAPIEntity oneWayFarePolicies
+--       }
 
 updateFarePolicy :: (HasCacheConfig r, EsqDBFlow m r, CoreMetrics m, HedisFlow m r) => SP.Person -> Id DFarePolicy.FarePolicy -> UpdateFarePolicyReq -> m UpdateFarePolicyRes
 updateFarePolicy admin fpId req = do
   runRequestValidation validateUpdateFarePolicyRequest req
   farePolicy <- SFarePolicy.findById fpId >>= fromMaybeM NoFarePolicy
-  unless (admin.merchantId == farePolicy.merchantId) $ throwError AccessDenied
+  -- unless (admin.merchantId == farePolicy.merchantId) $ throwError AccessDenied
```
Comment (DonDokidon) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/FarePolicy.hs:87]: Why?

Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/FarePolicy.hs:87]: Yes, because we dont have merchantId stored in farepolicy anymore. Actually i will comment this APIs as it is not used anymore

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/UI/FarePolicy.hs [hunk 5]
```diff
@@ -34,21 +33,22 @@ import Tools.Auth
 type API =
   "org"
```
[the remark this answers is not in the document]
Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/UI/FarePolicy.hs:34]: @hkmangla Should I delete the APIs only ? I had commented out some portions

File: Backend/dev/migrations/dynamic-offer-driver-app/0164-special-location-tag.sql (hunk 2 above)
Comment (khuzema786) [reply Backend/dev/migrations/dynamic-offer-driver-app/0164-special-location-tag.sql]: Done, added on special_location_priority and fare_product table

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FarePolicy.hs [hunk 6]
```diff
@@ -0,0 +1,40 @@
+module SharedLogic.FarePolicy where
```
[the remark this answers is not in the document]
Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FarePolicy.hs:10]: Done

**Review by DonDokidon — APPROVED** [review]

File: Backend/dev/migrations/dynamic-offer-driver-app/0172-special-location-tag.sql [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+-- Details : 4 Default Entries for 4 vehicle variants Per Merchant (Mandatory)
+INSERT INTO atlas_driver_offer_bpp.fare_product (id, merchant_id, fare_policy_id, vehicle_variant, "area", flow)
+    SELECT
+        md5(random()::text || clock_timestamp()::text)::uuid,
+        T1.merchant_id,
+        T1.id,
+        T1.vehicle_variant,
+        'Default',
+        'NORMAL'
+    FROM atlas_driver_offer_bpp.fare_policy AS T1 where T1.merchant_id = 'favorit0-0000-0000-0000-00000favorit';
```
Comment (hkmangla) [inline Backend/dev/migrations/dynamic-offer-driver-app/0172-special-location-tag.sql:60]: remove this filter on merchant_id, let backfill for all merchant

File: Backend/dev/migrations/dynamic-offer-driver-app/0172-special-location-tag.sql [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+CREATE TABLE atlas_driver_offer_bpp.special_location_priority (
+  id character(36) NOT NULL PRIMARY KEY,
+  merchant_id character(36) NOT NULL REFERENCES atlas_driver_offer_bpp.merchant (id),
+  category character varying(255) NOT NULL,
+  pickup_priority integer NOT NULL,
+  drop_priority integer NOT NULL
+);
+CREATE INDEX idx_special_location_priority ON atlas_driver_offer_bpp.special_location_priority USING btree (merchant_id, category);
+-- All Category Priorities for different supported locations based on MerchantId
+INSERT INTO atlas_driver_offer_bpp.special_location_priority VALUES ('f0e77e85-11fa-4949-af8e-92d8b915e3d9','favorit0-0000-0000-0000-00000favorit','SureMetro','2','2');
```
Comment (hkmangla) [inline Backend/dev/migrations/dynamic-offer-driver-app/0172-special-location-tag.sql:15]: Comment all this code and remove commented code. All this should go inside seed files

File: Backend/app/rider-platform/rider-app/Main/src/Beckn/ACL/OnSelect.hs [hunk 9]
```diff
@@ -85,6 +85,9 @@ buildQuoteInfo item = do
       estimatedTotalFare = roundToIntegral item.price.offered_value
       descriptions = item.quote_terms
   validatePrices estimatedFare estimatedTotalFare
+  specialLocationTag <- runMaybeT $ do
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/Beckn/ACL/OnSelect.hs:88]: `join . (.special_location_tag) <$> item.tags`

File: Backend/app/rider-platform/rider-app/Main/src/Beckn/ACL/OnSearch.hs [hunk 10]
```diff
@@ -101,7 +101,9 @@ buildEstimateOrQuoteInfo item = do
       waitingCharges = buildWaitingChargeInfo <$> item.tags
       driversLocation = fromMaybe [] $ item.tags <&> (.drivers_location)
   validatePrices estimatedFare estimatedTotalFare
-
+  specialLocationTag <- runMaybeT $ do
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/Beckn/ACL/OnSearch.hs:104]: join . (.special_location_tag) <$> item.tags

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/FareProduct.hs [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+ as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version. This program
+
+ is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+
+ or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details. You should have received a copy of
+
+ the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+-}
+
+module Storage.Queries.FareProduct where
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/FareProduct.hs:15]: Add these queries in cached Queries too and add warning on import of Storage queries

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FareProduct.hs [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+  deriving (Show, Read, Generic, Eq, ToSchema)
+
+data Area
+  = Pickup (Id SpecialLocation)
+  | Drop (Id SpecialLocation)
+  | Default
+  deriving (Generic, Eq, ToSchema)
+
+instance Read Area where
+  readsPrec _ ('P' : 'i' : 'c' : 'k' : 'u' : 'p' : '_' : uuid) = [(Pickup (Id $ T.pack uuid), "")]
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FareProduct.hs:37]: why this?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FareProduct.hs [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+  readsPrec _ ('D' : 'r' : 'o' : 'p' : '_' : uuid) = [(Pickup (Id $ T.pack uuid), "")]
+  readsPrec _ "Default" = [(Default, "")]
+  readsPrec _ _ = fail "Invalid format for Area"
+
+instance Show Area where
+  show (Pickup specialLocationId) = "Pickup_" <> T.unpack specialLocationId.getId
+  show (Drop specialLocationId) = "Drop_" <> T.unpack specialLocationId.getId
+  show Default = "Default"
+
+-- instance Read Area where
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FareProduct.hs:47]: remove commented code

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FarePolicy.hs [hunk 14]
```diff
@@ -73,33 +72,53 @@ getFarePolicyType farePolicy = case farePolicy.farePolicyDetails of
...
+
+farePolicyToFullFarePolicy :: Id Merchant -> Variant -> FarePolicy -> FullFarePolicy
+farePolicyToFullFarePolicy merchantId vehicleVariant FarePolicy {..} =
+  FullFarePolicy {..}
+
+-----------------------------------------------------------------------------------------------------------------------------------------
+------------------------------------------------APIEntity--------------------------------------------------------------------------------
+-----------------------------------------------------------------------------------------------------------------------------------------
+
+-- data FarePolicyAPIEntity = FarePolicyAPIEntity
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FarePolicy.hs:99]: remove

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/EndRide.hs [hunk 15]
```diff
@@ -115,7 +117,7 @@ buildEndRideHandle merchantId = do
       }
 
 driverEndRide ::
-  (MonadThrow m, Log m, MonadTime m, MonadGuid m) =>
+  (MonadFlow m, MonadReader r m, HasEsqEnv m r, CacheFlow m r) =>
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/EndRide.hs:120]: Why these changes?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/FarePolicy.hs [hunk 16]
```diff
@@ -74,19 +72,19 @@ validateUpdateFarePolicyRequest UpdateFarePolicyReq {..} =
...
     ]
 
-listFarePolicies :: (HasCacheConfig r, EsqDBFlow m r, HedisFlow m r) => SP.Person -> m ListFarePolicyRes
-listFarePolicies person = do
-  oneWayFarePolicies <- SFarePolicy.findAllByMerchantId person.merchantId
-  pure $
-    ListFarePolicyRes
-      { oneWayFarePolicies = map makeFarePolicyAPIEntity oneWayFarePolicies
-      }
+-- listFarePolicies :: (HasCacheConfig r, EsqDBFlow m r, HedisFlow m r) => SP.Person -> m ListFarePolicyRes
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/FarePolicy.hs:75]: remove

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/FarePolicy.hs [hunk 17]
```diff
@@ -13,10 +13,8 @@
 -}
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/FarePolicy.hs:13]: remove this all file.

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs [hunk 18]
```diff
@@ -131,52 +133,57 @@ handler merchant sReq = do
...
-              searchRequestSpecialZone
-              fareParams
-              merchant.id
-              result.distance
-              farePolicy.vehicleVariant
-              result.duration
-        Esq.runTransaction $
-          for_ listOfSpecialZoneQuotes QQuoteSpecialZone.create
-        return (Just (mkQuoteInfo fromLocation toLocation now <$> listOfSpecialZoneQuotes), Nothing)
+        flowType <- getFlowType merchantId fromLocationLatLong toLocationLatLong
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs:152]: `getFareProductForVariants` query is repeated in `getFlowType` and `getFarePolicyForVariants`, can we query once and use at same value at both the places ?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs [hunk 19]
```diff
@@ -131,52 +133,57 @@ handler merchant sReq = do
...
+  allFarePolicies <- getFarePolicyForVariants merchantId fromLocationLatLong toLocationLatLong
+  let farePolicies = selectFarePolicy result.distance allFarePolicies
   (quotes, mbEstimateInfos) <-
-    if isJust mbSpecialLocation
+    if isJust mbSpecialLocationTag
       then do
         whenJustM
           (QSearchRequestSpecialZone.findByMsgIdAndBapIdAndBppId sReq.messageId sReq.bapId merchantId)
           (\_ -> throwError $ InvalidRequest "Duplicate Search request")
         searchRequestSpecialZone <- buildSearchRequestSpecialZone sReq merchantId fromLocation toLocation result.distance result.duration
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs:149]: this is not required in case of `DFareProduct.NORMAL` flowType, right ? if so please move it under `DFareProduct.RIDE_OTP` case match

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs [hunk 20]
```diff
@@ -131,52 +133,57 @@ handler merchant sReq = do
...
+                  mbSpecialLocationTag
+            Esq.runTransaction $
+              for_ listOfSpecialZoneQuotes QQuoteSpecialZone.create
+            return (Just (mkQuoteInfo fromLocation toLocation now <$> listOfSpecialZoneQuotes), Nothing)
+          DFareProduct.NORMAL -> do
+            driverPoolCfg <- getDriverPoolConfig merchantId result.distance
+            estimateInfos <- buildEstimatesInfos fromLocation toLocation driverPoolCfg result farePolicies mbSpecialLocationTag
+            return (Nothing, Just estimateInfos)
       else do
         driverPoolCfg <- getDriverPoolConfig merchantId result.distance
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs:185]: can you avoid this repetition of code in NORMAL flow type and else part, both seems to be same only.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Select.hs [hunk 21]
```diff
@@ -62,7 +64,7 @@ handler :: DM.Merchant -> DSelectReq -> DEst.Estimate -> Flow ()
 handler merchant sReq estimate = do
   let merchantId = merchant.id
   searchReq <- QSR.findById estimate.requestId >>= fromMaybeM (SearchRequestNotFound estimate.requestId.getId)
-  farePolicy <- FarePolicyS.findByMerchantIdAndVariant merchant.id estimate.vehicleVariant >>= fromMaybeM NoFarePolicy
+  farePolicy <- getFarePolicyForVariant merchantId (LatLong searchReq.fromLocation.lat searchReq.fromLocation.lon) (LatLong searchReq.toLocation.lat searchReq.toLocation.lon) estimate.vehicleVariant
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Select.hs:67]: let's keep farePolicyId in the estimate/quote table ? that way we would not have to do call location service again and again in each API.
and here you will get the selected vahicle variant so in searchTry you can put the final farePolicyId that is to be used in subsequent APIs.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide/Internal.hs [hunk 22]
```diff
@@ -97,7 +98,7 @@ cancelRideImpl rideId bookingCReason = do
     transpConf <- QTC.findByMerchantId merchant.id >>= fromMaybeM (TransporterConfigNotFound merchant.id.getId)
     let searchRepeatLimit = transpConf.searchRepeatLimit
     now <- getCurrentTime
-    farePolicy <- QFP.findByMerchantIdAndVariant searchReq.providerId searchTry.vehicleVariant >>= fromMaybeM NoFarePolicy
+    farePolicy <- getFarePolicyForVariant searchReq.providerId (LatLong searchReq.fromLocation.lat searchReq.fromLocation.lon) (LatLong searchReq.toLocation.lat searchReq.toLocation.lon) searchTry.vehicleVariant
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide/Internal.hs:101]: this can be a point query directly to farePolicy table after adding farePolicyId to searchTry.

**Review by piyushKumar-1 — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs (hunk 19 above)
Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs:149]: No, use case is that for SpecialLocation...We can have normal estimate ride flow as well. Unlike Jatri Saathi usecase.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs (hunk 18 above)
Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs:152]: Yes but getFlowType is used only in one place, we dont want to compute it all other places where FarePolicy is required. One thing is that I always calculate FareProduct outside the query and then pass it but then whereever FarePolicy is required first i will have to always query FareProduct and pass it to get FarePolicy, I rather put all that logic for linking with FareProduct at one place.

Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs:152]: Is there some better way to do it, I dont want to compute FareProduct outside and pass it as an argument to `getFarePolicyForVariants`

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FarePolicy.hs [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+  fareProduct <- FareProduct.getFareProductForVariant merchantId fromlocaton toLocation vehVariant
+  farePolicy <- QFP.findById fareProduct.farePolicyId >>= fromMaybeM NoFarePolicy
+  return $ farePolicyToFullFarePolicy merchantId vehVariant farePolicy
+
+getFarePolicyForVariants :: (MonadFlow m, MonadReader r m, HasEsqEnv m r, CacheFlow m r) => Id Merchant -> LatLong -> LatLong -> m [FullFarePolicy]
+getFarePolicyForVariants merchantId fromlocaton toLocation = do
+  fareProducts <- FareProduct.getFareProductForVariants merchantId fromlocaton toLocation
+  mapM
+    ( \fareProduct -> do
+        farePolicy <- QFP.findById fareProduct.farePolicyId >>= fromMaybeM NoFarePolicy
```
[the remark this answers is not in the document]
Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FarePolicy.hs:44]: Here actually For each farepolicy I am associating vehicleVariant and merchantId from fareProduct. So will have to do it one by one and construct fullFarePolicy. As now farepolicy table we wont be storing the context of merchantId and vehicleVariant.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareProduct.hs [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+import qualified Storage.Queries.FareProduct as QFareProduct
+import Tools.Error
+
+getPickupSpecialLocation ::
+  (MonadFlow m, MonadReader r m, HasEsqEnv m r) =>
+  Id Merchant ->
+  DSpecialLocation.SpecialLocation ->
+  m (DSpecialLocation.SpecialLocation, Int)
+getPickupSpecialLocation merchantId pickupSpecialLocation = do
+  pickupSpecialLocationPriority <- QSpecialLocationPriority.findByMerchantIdAndCategory merchantId.getId pickupSpecialLocation.category
```
[the remark this answers is not in the document]
Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareProduct.hs:32]: Yes, Done

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs (hunk 20 above)
Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs:185]: Yes, Done.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/FareProduct.hs (hunk 11 above)
Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/FareProduct.hs:15]: Done

File: Backend/app/rider-platform/rider-app/Main/src/Beckn/ACL/OnSearch.hs (hunk 10 above)
Comment (khuzema786) [reply Backend/app/rider-platform/rider-app/Main/src/Beckn/ACL/OnSearch.hs:104]: Done

File: Backend/app/rider-platform/rider-app/Main/src/Beckn/ACL/OnSelect.hs (hunk 9 above)
Comment (khuzema786) [reply Backend/app/rider-platform/rider-app/Main/src/Beckn/ACL/OnSelect.hs:88]: Done

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FareProduct.hs (hunk 13 above)
Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/FareProduct.hs:47]: Done

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Select.hs (hunk 21 above)
Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Select.hs:67]: Can I do this as Next PR ?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide/Internal.hs (hunk 22 above)
Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide/Internal.hs:101]: Can I do this as Next PR ?

File: Backend/dev/migrations/dynamic-offer-driver-app/0172-special-location-tag.sql (hunk 8 above)
Comment (khuzema786) [reply Backend/dev/migrations/dynamic-offer-driver-app/0172-special-location-tag.sql]: I have removed hardcoded merchantId so now it's no longer a hardcoded migration that will only work with dev environment. Now it's fine right ?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs (hunk 18 above)
Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Search.hs:146]: Yes, Cool I handled this case, I have put flow in Full Fare Policy to avoid DB call.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide/Internal.hs (hunk 22 above)
Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide/Internal.hs]: Done Resolved.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Select.hs (hunk 21 above)
Comment (khuzema786) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Select.hs]: Resolved

**Review by piyushKumar-1 — APPROVED** [review]

**Review by hkmangla — CHANGES_REQUESTED** [review]

**Review by hkmangla — APPROVED** [review]
